### PR TITLE
python@3.12, python@3.11, python@3.10, python@3.9: Codesign Python.framework

### PR DESCRIPTION
**Pull request by AuroraWright**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----
As per https://github.com/Homebrew/brew/issues/16660 . I only updated currently supported Python versions, and I tried to follow the way it's done in other formulaes.

### CODE REVIEW

File: Formula/p/python@3.10.rb [hunk 1]
```diff
@@ -4,7 +4,7 @@ class PythonAT310 < Formula
   url "https://www.python.org/ftp/python/3.10.13/Python-3.10.13.tgz"
   sha256 "698ec55234c1363bd813b460ed53b0f108877c7a133d48bde9a50a1eb57b7e65"
   license "Python-2.0"
-  revision 2
+  revision 3
```
Comment (MikeMcQuaid) [inline Formula/p/python@3.10.rb:7]: This doesn't justify a `revision`, we'll distribute new bottles though.

File: Formula/p/python@3.10.rb [hunk 2]
```diff
@@ -256,6 +256,9 @@ def install
 
       # Remove symlinks that conflict with the main Python formula.
       rm %w[Headers Python Resources Versions/Current].map { |subdir| frameworks/"Python.framework"/subdir }
+
+      # Fix for https://github.com/Homebrew/brew/issues/16660
```
Comment (MikeMcQuaid) [inline Formula/p/python@3.10.rb:260]:
```suggestion
      # Binaries are codesigned but the framework is not
      # https://github.com/Homebrew/brew/issues/16660
```

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks for the PR @AuroraWright. Seems fine if it works as expected!

Some questions for you and @Homebrew/maintainers:
- should this change be requested in Python itself to be part of that build process?
- is there something special about Python here that means we should be signing this but not other Frameworks? Put another way: should Homebrew/brew be doing ad-hoc signing of all Frameworks like this?

**Comment by AuroraWright** [comment]
> Thanks for the PR @AuroraWright. Seems fine if it works as expected!
> 
> Some questions for you and @Homebrew/maintainers:
> 
> * should this change be requested in Python itself to be part of that build process?
> * is there something special about Python here that means we should be signing this but not other Frameworks? Put another way: should Homebrew/brew be doing ad-hoc signing of all Frameworks like this?

Python already seems to codesign the framework in the official releases (not sure when/how they do that).
I'm not sure if it should be extended to any framework, but maybe if there are no side effects (can't think of any myself) it would be a good idea if Apple relies on it to add access to some OS APIs like this case has shown

**Comment by fxcoudert** [comment]
I'm skeptical that this is the right approach, and that it actually works. This is signing the frameworks during build, but won't the signature be invalidated during bottling & bottle pouring? This is the reason why we resign binaries in `brew` when we pour them.

Also, if Python expects frameworks to be signed, I think the build process should sign them. It's an ad hoc signature, so it's easy to do. Are the frameworks shipped by upstream (python.org) actually ad-hoc signed, or signed with a developer certificate?

**Comment by MikeMcQuaid** [comment]
> This is signing the frameworks during build, but won't the signature be invalidated during bottling & bottle pouring?

Yes, seems likely unless we resign with another ad-hoc certificate.

> Are the frameworks shipped by upstream (python.org) actually ad-hoc signed, or signed with a developer certificate?

@AuroraWright in case it helps: the question behind this question is: unless they are also ad-hoc signed upstream, unfortunately nothing we can do in Homebrew will address this properly because we don't sign anything currently.

**Comment by AuroraWright** [comment]
> > Are the frameworks shipped by upstream (python.org) actually ad-hoc signed, or signed with a developer certificate?
> 
> @AuroraWright in case it helps: the question behind this question is: unless they are also ad-hoc signed upstream, unfortunately nothing we can do in Homebrew will address this properly because we don't sign anything currently.
Everything they distribute (binaries, framework, etc) is signed with their developer certificate. But, to be honest, I'm not sure why it makes a difference here since (as far as I know) adhoc vs dev certificate only matters to make gatekeeper complain less with you launch a GUI app and it shouldn't make a difference accessing APIs and the like (except if you need some entitlement).

**Pull request closed without merging**